### PR TITLE
[BUG] Error cutting decimals in yaml.dump > ruamel.yaml.YAML().dump()

### PR DESCRIPTION
**Pull request by GuillermoAbadLopez**
## PR using `ruamel.yaml.YAML().dump()` as standard yaml dumping in qililab to not lose decimals, and work with scientific notation (1e-9)!

Since we want to use scientific notation and large decimals, I've added a test to check the serialization works with:
- large decimals
- scientific notation 1e-9
- underscore notaiton 100_000_000
- None's

## Problems/Warnings I see:
- `ruamel.yaml.YAML().dump` needs a stream almost always, so I had to pass `io.BytesIO()` in various places where before `io` was not being used.

- I see Albert had a yaml_representer for **Null values**, aswell, I dunno if those are supported by `ruamel.yaml`, but I have searched and they are not used in the runcards I have, only None's are being used, and those work!

- Its a longer and uglier code, maybe we could define a  `QililabYAML = ruamel.yaml.YAML()` ?? Not necessary though

- If something was previously in scientific notation or in under score notation, they get translated to normal float for the next Runcard. So Runcard don't keep exactly the format they are written with, but everything works together. Also I guess this already happened before, so not really a problem.

## Other problem we are gonna solve in this PR:
- Fixing the version of `ruamel` in the requirements.txt so main doesn't fail in the future.

### CODE REVIEW

File: src/qililab/experiment/base_experiment.py
Comment (AlbertMitjans) [inline src/qililab/experiment/base_experiment.py:24]: Instead of using this import and then `ruamel.yaml.YAML` we could import the YAML class directly:
```suggestion
from ruamel.yaml import YAML
```

**Review by AlbertMitjans — COMMENTED** [review]
Looking at the `ruamel` documentation, I don't think we need a context manager anymore 🤔

File: src/qililab/experiment/base_experiment.py
Comment (GuillermoAbadLopez) [reply src/qililab/experiment/base_experiment.py:24]: Indeed! It will look way better that way yes!

**Comment by GuillermoAbadLopez** [comment]
Yeah, I'll leave that to you, since I am not very familiar with it! I guess you are talking about how we can pass the io.ByTestIO() instead than using the context manager?

Another thing, from what I've seen `ruamel` has a very broad acceptation as an improvement for the basic PyYAML!

File: src/qililab/experiment/base_experiment.py
Comment (GuillermoAbadLopez) [reply src/qililab/experiment/base_experiment.py]: Quickly applied this change!

File: src/qililab/experiment/base_experiment.py
Comment (AlbertMitjans) [inline src/qililab/experiment/base_experiment.py:130]: Can we use `sort_keys=False` with ruamel to make sure the functionality is the same?

File: src/qililab/experiment/base_experiment.py
Comment (AlbertMitjans) [inline src/qililab/experiment/base_experiment.py:360]: Same here

File: src/qililab/experiment/base_experiment.py
Comment (AlbertMitjans) [inline src/qililab/experiment/base_experiment.py:417]: Same here

File: src/qililab/instrument_controllers/instrument_controllers.py
Comment (AlbertMitjans) [inline src/qililab/instrument_controllers/instrument_controllers.py:68]: What is this new addition?

File: tests/test_data_management.py
Comment (AlbertMitjans) [inline tests/test_data_management.py:164]: Tests shouldn't create files! We should remove this.

Comment (AlbertMitjans) [reply tests/test_data_management.py:164]: There should be a way of channeling `yaml.dump` into a variable instead of a file

**Review by AlbertMitjans — COMMENTED** [review]
Added some comments!

File: src/qililab/experiment/base_experiment.py
Comment (GuillermoAbadLopez) [reply src/qililab/experiment/base_experiment.py:130]: In the same post, you can see that there is actually a `.sort()` method, but for the class `YAML()`, so you first would need to have the yaml in such class, order it as you want with lambas, and then dump it.

But as said before, I think we don't need this, since the Default way already conserves the order!

The only problem we would have would be, in the opposite ways, when we actually want a alphabetical order, but I don't think we want that anywhere, no?

File: src/qililab/experiment/base_experiment.py
Comment (GuillermoAbadLopez) [reply src/qililab/experiment/base_experiment.py:360]: answered above

File: src/qililab/experiment/base_experiment.py
Comment (GuillermoAbadLopez) [reply src/qililab/experiment/base_experiment.py:417]: answered above

File: src/qililab/experiment/base_experiment.py
Comment (GuillermoAbadLopez) [reply src/qililab/experiment/base_experiment.py:130]: Did this in my terminal to check it:

File: src/qililab/instrument_controllers/instrument_controllers.py
Comment (GuillermoAbadLopez) [reply src/qililab/instrument_controllers/instrument_controllers.py:68]: `ruamel`, always needs a `stream`, unless you call it from a `context_manager`, so I just added default `stream` from `io` which I found in a stack exchange. 

`ruamel` also says this in their documentation, we could maybe use that:
Screenshot 2023-10-25 at 19 28 11

File: tests/test_data_management.py
Comment (GuillermoAbadLopez) [reply tests/test_data_management.py]: What about removing the file in the end, as I just did, with `os.remove(path)`?

I've seen this from some other parts in qililab, is that acceptable?

File: src/qililab/experiment/base_experiment.py
Comment (AlbertMitjans) [reply src/qililab/experiment/base_experiment.py:130]: We don't want to sort anything, thus we are good! Good job! 🚀 💪

File: src/qililab/instrument_controllers/instrument_controllers.py
Comment (AlbertMitjans) [reply src/qililab/instrument_controllers/instrument_controllers.py:68]: Great!

File: tests/test_data_management.py
Comment (AlbertMitjans) [reply tests/test_data_management.py]: I prefer not do it. I think you already had a bad experience with tests creating files 😆

Comment (AlbertMitjans) [reply tests/test_data_management.py]: You can add a stream like you did in other cases!!

**Review by AlbertMitjans — COMMENTED** [review]
Good job! I will approve when the test stops creating files!

File: tests/test_data_management.py
Comment (GuillermoAbadLopez) [reply tests/test_data_management.py]: Hahha you convinces me already with the first statement 🤣

File: tests/test_data_management.py
Comment (AlbertMitjans) [inline tests/test_data_management.py:166]: Instead of creating and removing a file, why not using `YAML().dump` to dump the information into a stream, and then retrieve it from there?

[See this link for more information.](https://stackoverflow.com/questions/61722242/dump-the-yaml-to-a-variable-instead-of-streaming-it-in-stdout-using-ruamel-yaml)

Comment (GuillermoAbadLopez) [reply tests/test_data_management.py:166]: Yess, that what I said yesterday that I was gonna do! Thanks!!!

I think I have actually done that already above in this discussion, in the console image I pasted! 🙌

Comment (GuillermoAbadLopez) [reply tests/test_data_management.py:166]: @AlbertMitjans I don't see the solution that straight forward, now that I have looked it better... 😬 

Because I wanna test this function:
Screenshot 2023-10-27 at 21 03 15
so I would need to patch in some way the with open, so that it doesn't open a file and saves it into a variable...

Notice I cannot just do what you said, because I would then just not be testing this function that we wanna test 😅

Comment (AlbertMitjans) [reply tests/test_data_management.py:166]: The `save_platform` function is already tested in this file. As you mention in the comments, the tests you added are to make sure ruamel works properly. Thus I see no need of using `save_platform`.

I would suggest using `YAML().dump` directly to obtain the `saved_platform` variable, and then run the tests.

Comment (GuillermoAbadLopez) [reply tests/test_data_management.py:166]: I mean, it makes sense, but I don't know if I completely agree on that. 🤔 

Since the idea I had for this test, was testing that both methods work together properly, kind of like an integration test of both functions. 

What you are proposing is an extended unit test for the `build_platform`, working correctly with the YAML dump function. 

And if in the future someone changes the `save_platform` but not the `build_platform` one, then the test would still pass, when maybe it shouldn't? 🤔  

What do you think? Still, think it's better to not test the integration of both methods?

Comment (GuillermoAbadLopez) [reply tests/test_data_management.py:166]: Since this conversation is going too long, for a single test, I'll do as you say in the next answer hehe 🙃

File: tests/test_data_management.py
Comment (AlbertMitjans) [inline tests/test_data_management.py:209]: By the way, could we merge all of these tests in a single one?

Just doing the following assert would suffice, right?
```python
assert original_platform == saved_platform
```

File: tests/test_data_management.py
Comment (AlbertMitjans) [reply tests/test_data_management.py:166]: Check my new comment. If you want to test that `save_platform` saves the exact same dictionary, I believe one test would suffice, meaning:

```python
original_platform = ql.build_platform(Galadriel.runcard)
path = save_platform(path="./test.yml", platform=original_platform)
saved_platform = ql.build_platform(path)
os.remove(path)
assert original_platform == saved_platform
```

I would still do this test without using `save_platform`:

```python
original_platform = Galadriel.runcard
saved_platform = YAML().dump(original_platform.to_dict(), stream=...)
assert original_platform == saved_platform
```

Comment (GuillermoAbadLopez) [reply tests/test_data_management.py:166]: Bueno, en el primer caldria possar un `to_dict` por ahi, porque sino les dues classes de Platform no seran les mismes

Comment (jjmartinezQT) [reply tests/test_data_management.py:166]: Please Guille, use english so everyone in the team can follow up on these discussions.

Comment (GuillermoAbadLopez) [reply tests/test_data_management.py:166]: Yesyes, I actually, had deleted the comment already!

File: tests/test_data_management.py
Comment (GuillermoAbadLopez) [reply tests/test_data_management.py]: Reduced them by joining them with the already existing tests!

File: tests/test_data_management.py
Comment (GuillermoAbadLopez) [reply tests/test_data_management.py]: In the end, as we commented, I ended up reducing the tests, by joining them with the already existing one very in few lines 👍

File: tests/test_data_management.py
Comment (AlbertMitjans) [inline tests/test_data_management.py:95]: Just in case, can you do a `copy.deepcopy` of this dictionary? To make sure the changes below don't affect other tests.

File: tests/test_data_management.py
Comment (AlbertMitjans) [inline tests/test_data_management.py:101]: This file doesn't exist, how is this even working?

Comment (AlbertMitjans) [reply tests/test_data_management.py:101]: Same thing for the test below

File: tests/test_data_management.py
Comment (AlbertMitjans) [inline tests/test_data_management.py:127]: Why did you remove this?

**Review by AlbertMitjans — COMMENTED** [review]
Added some comments!

File: tests/test_data_management.py
Comment (GuillermoAbadLopez) [reply tests/test_data_management.py:95]: Agree, I'll do that!

File: tests/test_data_management.py
Comment (GuillermoAbadLopez) [reply tests/test_data_management.py:101]: What do you mean it doesn't exist? It is there: 
https://github.com/qilimanjaro-tech/qililab/blob/main/examples/runcards/galadriel.yml

File: tests/test_data_management.py
Comment (GuillermoAbadLopez) [reply tests/test_data_management.py:127]: Explained in the answer above!

File: tests/test_data_management.py
Comment (GuillermoAbadLopez) [reply tests/test_data_management.py:101]: The thing is, in this test we were testing `Galadriel.runcard` from `data.py`, and we just also read the `yaml` in `galadriel.yaml`, to then check that they are not the same... but that doesn't make any sense to test..

Comment (AlbertMitjans) [reply tests/test_data_management.py:101]: Ok my bad.

Comment (AlbertMitjans) [reply tests/test_data_management.py:101]: I don't like at all using files outside the `test` folder inside tests. Good thing you deleted it

Comment (AlbertMitjans) [reply tests/test_data_management.py:101]: > Like why would they be different? I mean they are, because they are to different dictionaries, but it doesn't;t make sense to add here a line for the sake of checking that, since that is no functionality at all 🙃

Completely agree!

File: tests/test_data_management.py
Comment (AlbertMitjans) [reply tests/test_data_management.py:127]: Totally agree!

**Review by AlbertMitjans — APPROVED** [review]
Good job!! 💪

**Pull request merged**
